kiwi/markup/__init__.py

```diff
@@ -44,8 +44,11 @@ def new(description: str, name: str='any'):  # noqa: E252
             markup = Markup._load_markup_by_name(name, description)
             log.info('Support for multiple markup descriptions available')
         except KiwiAnyMarkupPluginError:
+            log.warn('Support for non-XML formatted config files requires '
+                     'the Python anymarkup module.')
+
```

Comment:
The warning message makes sense. I however fear this message could be overlooked with the following error information when we try to read in XML and it's not XML. I thought in addition to the warning we could try to read in the data as XML if the description file is **not** named ```config.xml``` but ```something.kiwi```. If the XML reading fails we just fail with a message that the given file ```something.kiwi``` doesn't look like an XML document.

So just the etree.parse() step without the schema and schematron validation. This will fail only for invalid XML
Thoughts ?

Reply:
Yeah if we are happy to attempt an etree.parse() just to see if it is valid XML at this point we can do that. I didn't want to rush into doing that as past experience has been that that can be an expensive operation. ;-)

However should we really be coding specific file name handling in here at this point? Would it be better to just keep the checking generic, and check explicitly for the referenced file being parsible as XML, and if not generate the above message as an error rather than a warning?

Reply:
I agree it's probably a bad idea. Ok, let's keep things as you have done it

Reply:
I gave it a short test and of course it works, I can see the message but also lots of other errors when you pass in yaml when we need XML. There is the following code in kiwi/kiwi/xml_description.py::load()

```
try:
    description = etree.parse(self.description)
    validation_rng = relaxng.validate(description)
    validation_schematron = schematron.validate(description)
except Exception as issue:
    ...
```

as you can see it raises on any error that comes out of etree.parse. I was thinking if we might raise on different exception types that comes from the parser. I think there is one for invalid XML and instead of bumping the whole chain of error information we could condense the information a bit such that your warning message is also in the eye if the user. Thoughts ?

Reply:
Sorry if I sound picky on this. The original idea was to give the user a better "error" impression when yaml is passed and no anymarkup exists. With the additional warning we created more information and I fear users will be overwhelmed by the amount of "stuff" that is presented to them in this case

Reply:
This is what you get right now

```
[ INFO    ]: 20:29:43 | Support for XML markup available
[ ERROR   ]: 20:29:43 | Unexpected error:
Traceback (most recent call last):
  File "/home/ms/kiwi/.tox/3.6/bin/kiwi-ng", line 33, in <module>
    sys.exit(load_entry_point('kiwi', 'console_scripts', 'kiwi-ng')())
  File "/home/ms/kiwi/kiwi/kiwi.py", line 68, in main
    App()
  File "/home/ms/kiwi/kiwi/app.py", line 35, in __init__
    task_class().process()
  File "/home/ms/kiwi/kiwi/tasks/image_info.py", line 73, in process
    self.command_args['--description']
  File "/home/ms/kiwi/kiwi/tasks/base.py", line 150, in load_xml_description
    config_file
  File "/home/ms/kiwi/kiwi/xml_description.py", line 70, in __init__
    self.description = self.markup.get_xml_description()
  File "/home/ms/kiwi/kiwi/markup/xml.py", line 37, in get_xml_description
    self.description
  File "/home/ms/kiwi/kiwi/markup/base.py", line 58, in apply_xslt_stylesheets
    etree.tostring(xslt_transform(etree.parse(description)))
  File "src/lxml/etree.pyx", line 3521, in lxml.etree.parse
  File "src/lxml/parser.pxi", line 1859, in lxml.etree._parseDocument
  File "src/lxml/parser.pxi", line 1885, in lxml.etree._parseDocumentFromURL
  File "src/lxml/parser.pxi", line 1789, in lxml.etree._parseDocFromFile
  File "src/lxml/parser.pxi", line 1177, in lxml.etree._BaseParser._parseDocFromFile
  File "src/lxml/parser.pxi", line 615, in lxml.etree._ParserContext._handleParseResultDoc
  File "src/lxml/parser.pxi", line 725, in lxml.etree._handleParseResult
  File "src/lxml/parser.pxi", line 654, in lxml.etree._raiseParseError
  File "/home/ms/Project/kiwi-descriptions/suse/x86_64/suse-tumbleweed-docker/config.kiwi", line 1
lxml.etree.XMLSyntaxError: Start tag expected, '<' not found, line 1, column 1

```

So I think we need to catch the error condition earlier, as it causes a stack trace on the xslt transformation

Reply:
So maybe it's enough to catch potential parse exceptions in apply_xslt_stylesheets() in addition to your warning message.

Reply:
So we could catch on lxml.etree.XMLSyntaxError, show the error and your warning

Reply:
You are not being picky.

The initial version was a first pass to get feedback as to whether I was taking the right approach or not. And the feedback was exactly the sort of thing I was looking for.

I've pushed up an alternate version using etree.parse() to check if the file is XML or not, catching etree.XMLSyntaxError, and generating an meaningful error when a YAML or JSON file is found for config.kiwi:

```
(kiwi_broken) fifteen[fergal] OSinside/kiwi % cat /tmp/blah/config.kiwi
{}
(kiwi_broken) fifteen[fergal] OSinside/kiwi % kiwi-ng image info --description /tmp/blah --print-xml
[ INFO    ]: 14:36:55 | Reading runtime config file: '/etc/kiwi.yml'
[ INFO    ]: 14:36:55 | Loading XML description
[ ERROR   ]: 14:36:55 | KiwiConfigFileFormatNotSupported: Support for non-XML formatted config files requires the Python anymarkup module.
(kiwi_broken) fifteen[fergal] OSinside/kiwi (1) % mv /tmp/blah/config.kiwi /tmp/blah/config.json 
(kiwi_broken) fifteen[fergal] OSinside/kiwi % mv /tmp/blah/config.yml /tmp/blah/config.kiwi 
(kiwi_broken) fifteen[fergal] OSinside/kiwi % cat /tmp/blah/config.kiwi                             
image:
  '@name': Tumbleweed-container
  '@schemaversion': '7.3'
  description:
    '@type': system
    author: David Cassany
    contact: dcassany@suse.com
    specification: openSUSE Tumbleweed container
  packages:
    '@patternType': plusRecommended
    '@type': bootstrap
    package:
    - '@name': patterns-openSUSE-base
    - '@name': ca-certificates
    - '@name': ca-certificates-mozilla
  preferences:
    packagemanager: zypper
    rpm-check-signatures: 'false'
    rpm-excludedocs: 'true'
    type:
      '@image': docker
      containerconfig:
        '@maintainer': David Cassany <dcassany@suse.com>
        '@name': opensuse
        '@tag': tumbleweed
    version: 1.0.4
  repository:
    '@alias': Tumbleweed
    source:
      '@path': obs://openSUSE:Tumbleweed/standard
(kiwi_broken) fifteen[fergal] OSinside/kiwi % kiwi-ng image info --description /tmp/blah --print-xml
[ INFO    ]: 14:38:11 | Reading runtime config file: '/etc/kiwi.yml'
[ INFO    ]: 14:38:11 | Loading XML description
[ ERROR   ]: 14:38:11 | KiwiConfigFileFormatNotSupported: Support for non-XML formatted config files requires the Python anymarkup module.
(kiwi_broken) fifteen[fergal] OSinside/kiwi (1) %
```
While things still work as before when anymarkup is available:
```
(kiwi) fifteen[fergal] OSinside/kiwi % kiwi-ng image info --description /tmp/blah --print-xml
[ INFO    ]: 14:38:50 | Reading runtime config file: '/etc/kiwi.yml'
[ INFO    ]: 14:38:50 | Loading XML description
[ INFO    ]: 14:38:50 | Support for multiple markup descriptions available
[ INFO    ]: 14:38:50 | --> loaded /tmp/blah/config.kiwi
[ INFO    ]: 14:38:50 | --> Selected build type: docker
{
    "image": "Tumbleweed-container"
}
[ INFO    ]: 14:38:50 | Description(XML):
<image name="Tumbleweed-container" schemaversion="7.3">
    <description type="system">
        <author>David Cassany</author>
        <contact>dcassany@suse.com</contact>
        <specification>openSUSE Tumbleweed container</specification>
    </description>
    <packages patternType="plusRecommended" type="bootstrap">
        <package name="patterns-openSUSE-base"/>
        <package name="ca-certificates"/>
        <package name="ca-certificates-mozilla"/>
    </packages>
    <preferences>
        <packagemanager>zypper</packagemanager>
        <rpm-check-signatures>false</rpm-check-signatures>
        <rpm-excludedocs>true</rpm-excludedocs>
        <type image="docker">
            <containerconfig maintainer="David Cassany &lt;dcassany@suse.com&gt;" name="opensuse" tag="tumbleweed"/>
        </type>
        <version>1.0.4</version>
    </preferences>
    <repository alias="Tumbleweed">
        <source path="obs://openSUSE:Tumbleweed/standard"/>
    </repository>
</image>
(kiwi) fifteen[fergal] OSinside/kiwi % 
```

Let me know how that looks, and, in particular, if the test I added is sufficient?